pkg/querier/blocks_store_queryable.go

```diff
@@ -986,7 +938,92 @@ func (q *blocksStoreQuerier) fetchSeriesFromStores(ctx context.Context, sp *stor
+	resp, err := stream.Recv()
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			return mySeries, myWarnings, myQueriedBlocks, myStreamingSeriesLabels, indexBytesFetched, false, false, err
+		}
+
+		if shouldRetry(err) {
+			return mySeries, myWarnings, myQueriedBlocks, myStreamingSeriesLabels, indexBytesFetched, false, true, nil
+		}
+
+		return mySeries, myWarnings, myQueriedBlocks, myStreamingSeriesLabels, indexBytesFetched, false, false, err
+	}
+	defer resp.FreeBuffer()
+
+	// Response may either contain series, streaming series, warning or hints.
+	if s := resp.GetSeries(); s != nil {
+		mySeries = append(mySeries, s.CloneRefs())
+
+		// Add series fingerprint to query limiter; will return error if we are over the limit
+		if err := queryLimiter.AddSeries(mimirpb.FromLabelAdaptersToLabels(s.Labels)); err != nil {
+			return mySeries, myWarnings, myQueriedBlocks, myStreamingSeriesLabels, indexBytesFetched, false, false, err
+		}
+
+		chunksCount, chunksSize := countChunksAndBytes(s)
+		q.metrics.chunksTotal.Add(float64(chunksCount))
+		if err := queryLimiter.AddChunkBytes(chunksSize); err != nil {
+			return mySeries, myWarnings, myQueriedBlocks, myStreamingSeriesLabels, indexBytesFetched, false, false, err
+		}
+		if err := queryLimiter.AddChunks(chunksCount); err != nil {
+			return mySeries, myWarnings, myQueriedBlocks, myStreamingSeriesLabels, indexBytesFetched, false, false, err
+		}
+		if err := queryLimiter.AddEstimatedChunks(chunksCount); err != nil {
+			return mySeries, myWarnings, myQueriedBlocks, myStreamingSeriesLabels, indexBytesFetched, false, false, err
+		}
+	}
+
+	if w := resp.GetWarning(); w != "" {
+		myWarnings.Add(errors.New(w))
+	}
+
+	if h := resp.GetHints(); h != nil {
+		hints := hintspb.SeriesResponseHints{}
+		if err := types.UnmarshalAny(h, &hints); err != nil {
+			return mySeries, myWarnings, myQueriedBlocks, myStreamingSeriesLabels, indexBytesFetched, false, false, errors.Wrapf(err, "failed to unmarshal series hints from %s", c.RemoteAddress())
+		}
+
+		ids, err := convertBlockHintsToULIDs(hints.QueriedBlocks)
+		if err != nil {
+			return mySeries, myWarnings, myQueriedBlocks, myStreamingSeriesLabels, indexBytesFetched, false, false, errors.Wrapf(err, "failed to parse queried block IDs from received hints")
+		}
+
+		myQueriedBlocks = append(myQueriedBlocks, ids...)
+	}
+
+	if s := resp.GetStats(); s != nil {
+		indexBytesFetched += s.FetchedIndexBytes
+	}
+
+	if ss := resp.GetStreamingSeries(); ss != nil {
+		myStreamingSeriesLabels = slices.Grow(myStreamingSeriesLabels, len(ss.Series))
+
+		var (
+			labelsBuilder labels.ScratchBuilder
+			ls            labels.Labels
+		)
+		for _, s := range ss.Series {
+			ls = mimirpb.FromLabelAdaptersOverwriteLabelsSafe(&labelsBuilder, s.Labels, &ls)
```

Comment:
Same as above - I think we can use `FromLabelAdaptersToLabelsWithCopy` here which should have the same performance characteristics as `FromLabelAdaptersToLabels` with `stringlabels`.

This removes the need for `FromLabelAdaptersOverwriteLabelsSafe` entirely.

Reply:
Done.